### PR TITLE
Fix: `torch.compile` was a no-op in `load_checkpoint`

### LINKED ISSUE #457: `torch.compile(self)` is a no-op — compiling `forward` directly gives 1.48x on A100
I think I found a bug worth flagging. I noticed the model never actually gets compiled despite torch_compile=True. Traced it back to this:

`self = torch.compile(self)` in `TimesFM_2p5_200M_torch_module.load_checkpoint` doesn't do anything. `self` inside a method is a local variable : reassigning it doesn't affect the reference.

```python
# timesfm_2p5_torch.py:84–89
if torch_compile:
    logging.info("Compiling model...")
    self = torch.compile(self)   # no-op: rebinds local only
```

I comfirmed it with one line after loading:

```python
model = timesfm.TimesFM_2p5_200M_torch.from_pretrained("google/timesfm-2.5-200m-pytorch")
print(type(model.model))
# TimesFM_2p5_200M_torch_module — never OptimizedModule
```

`torch_compile=True` (the default) has no effect. The log says "Compiling model..." but nothing gets compiled.

---

**I dug into whether fixing the no-op would actually help**

The KV cache writes in `MultiHeadAttention.forward` cause a graph break:

```python
start = decode_cache.next_index[0]
end = start + n_patches
decode_cache.key[:, start:end] = key  # ← break here
```

`torch._dynamo.explain(model.decode)` on the real model shows 10 subgraphs and 3 breaks, all at `transformer.py:279`. `capture_scalar_outputs=True` doesn't fix it. Dynamo doesn't recompile per step which is nice, but the cache write runs in eager on every AR step across all layers.

So even with `torch.compile` correctly applied to the full module, you get nothing on A100 with the real 200M model:

```
baseline (no-op)        108.8ms   1.00x
fixed (OptimizedModule) 109.1ms   1.00x
```

---

**Compiling `forward` directly does give a real speedup though** forward still graph-breaks at the KV-cache write (`decode_cache.key[:,start:end] = key`) once per layer. But those breaks are on cheap cache copies; the attention/MLP matmuls around them compile fine, so it still nets a real speedup. I tried compiling it directly:

```python
self.model.forward = torch.compile(self.model.forward)
```

A100, real model, batch=64, context=512, horizon=128:

```
eager                   108.8ms   1.00x
torch.compile(module)   109.1ms   1.00x
torch.compile(forward)   73.7ms   1.48x
```

For reference, `forward` alone gets 3.84x, the gap to 1.48x is the AR loop staying in Python. Still 48% faster end-to-end from a one-liner, feels worth it...

---

I'm not familiar enough with the codebase to know if there is a reason this wasn't done already. Would love to know if this direction makes sense. Happy to help if it does. I'm about to open a PR.

### PR DESCRIPTION
**Pull request by matdou**
Closes #457

`TimesFM_2p5_200M_torch_module.load_checkpoint` called `self = torch.compile(self)` inside the method body. That rebinds the local name `self` but doesn't mutate the object — the compiled version is immediately thrown away. Users passing `torch_compile=True` got no speedup at all.

The fix moves the compile call to the outer `TimesFM_2p5_200M_torch.load_checkpoint`, targeting `self.model.forward` directly:

```python
torch_compile = kwargs.pop("torch_compile", self.torch_compile)
self.model.load_checkpoint(model_file_path, **kwargs)
if torch_compile:
    self.model.forward = torch.compile(self.model.forward)
```

`nn.Module.__call__` dispatches through `self.forward`, so reassigning the instance attribute is enough for the compiled version to be used by the `decode()` path.

## Why it still helps despite graph breaks

`forward` does not trace into a single clean graph. Each transformer layer writes to its KV cache with a data-dependent slice:

```python
start = decode_cache.next_index[0]    # tensor value, needs an implicit .item()
decode_cache.key[:, start:end] = key  # transformer.py:279
```

Dynamo can't trace a slice whose index is a runtime tensor value, so it graph-breaks here — once per layer. `torch._dynamo.explain()` on the real decode call path (forward + a populated cache) shows **8 graphs / 7 breaks**.

The breaks land on cheap memory copies. The expensive work — the QKV/attention/MLP matmuls — sits inside the compiled subgraphs and gets fused and optimized normally. So compile still pays off, just not as much as a fully unbroken graph would.

Measured on A100 with `google/timesfm-2.5-200m-pytorch`, batch=64, context=512, horizon=128: **1.55x median speedup** (**111ms → 72ms**). Numerical diff vs uncompiled: max abs 7.2e-05 (TF32 rounding).

## Tests

Two new tests in `tests/test_model_loading.py` verify that `model.forward` is compiled after `load_checkpoint(torch_compile=True)` and left as a plain bound method when `torch_compile=False`.

## **Possible follow-up** (not in this PR)

The per-layer graph break could be removed by making the cache-write index non-data-dependent (e.g. `capture_scalar_outputs=True`, or an `index_copy`/`slice_scatter` with a precomputed index). That would unlock the remaining headroom but is a larger, separate change.

### CODE REVIEW

**Comment by matdou** [comment]
Benchmark

Run on an A100, `torch_compile=True`, batch=64, context=512, horizon=128:

```python
import time, torch, numpy as np
import timesfm
from timesfm import configs

torch.set_float32_matmul_precision("high")

HORIZON, CONTEXT, N_SERIES = 128, 512, 64
WARMUP, REPEATS = 3, 10

fc = configs.ForecastConfig(
    max_context=CONTEXT, max_horizon=HORIZON,
    normalize_inputs=True, per_core_batch_size=N_SERIES,
)
inputs = [np.sin(np.linspace(0, 20, CONTEXT) + i * 0.3).astype(np.float32)
          for i in range(N_SERIES)]

def bench(model):
    model.compile(fc)
    for _ in range(WARMUP):
        model.forecast(horizon=HORIZON, inputs=inputs)
    torch.cuda.synchronize()
    elapsed = []
    for _ in range(REPEATS):
        torch.cuda.synchronize()
        t0 = time.perf_counter()
        model.forecast(horizon=HORIZON, inputs=inputs)
        torch.cuda.synchronize()
        elapsed.append(time.perf_counter() - t0)
    return sorted(elapsed)[len(elapsed) // 2]

model = timesfm.TimesFM_2p5_200M_torch.from_pretrained(
    "google/timesfm-2.5-200m-pytorch", torch_compile=True,
)
print(f"median = {bench(model) * 1000:.1f} ms")
```

Same script, only difference is the installed `timesfm` version:

- `main`:    111.7 ms median
- this PR:    72.2 ms median  ->  **1.55x**

Output diff vs uncompiled is max abs 7.2e-05 (TF32 rounding).

**Comment by matdou** [comment]
cc @rajatsen91 @siriuz42, would appreciate a sanity check here when you have a moment.

**Comment by matdou** [comment]
Little update : I checked whether removing the per-layer graph break helps.

Swapping the data-dependent cache-write slice (`decode_cache.key[:, start:end] = key`) for an `index_copy_` with a precomputed index makes `forward` trace as one graph (0 breaks, down from 7), output bit-identical.

Then I benchmarked it and... ~~womp womp~~ : A100, same setup:

```
compile forward (this PR)     72.2 ms
+ break removed (index_copy_) 71.8 ms   1.01x
```

The breaks sit on cheap KV-cache copies, not the matmuls. The remaining cost is the autoregressive loop staying in Python, which this doesn't touch. Leaving it out to keep things lean, but I've got the branch if anyone wants to go pick a fight with the loop.

**Comment by rajatsen91** [comment]
I will take a look at the code shortly, but can you also test on horizons > 128. That is where the AR step kicks in. Horizon <= 128 is basically just the forward call.

**Comment by matdou** [comment]
@rajatsen91 I did : 

Results at context=512, A100 (WARMUP=10, REPEATS=20, median CUDA event timing):

| shape                  | eager ms | compiled ms | speedup |
|------------------------|----------|-------------|---------|
| b64/h128 (no AR)       | 108      | 76          | 1.43x   |
| b64/h256 (1 AR step)   | 202      | 137         | 1.48x   |
| b64/h512 (3 AR steps)  | 386      | 257         | 1.50x   |
| b64/h1024 (7 AR steps) | 749      | 489         | 1.53x   |

The speedup holds and slightly increases with horizon, each AR step is another compiled forward call, so the savings accumulate. The Python loop overhead is negligible relative to per-step compute (~91ms/step eager, ~70ms compiled across the h256→h1024 range).

**Review by rajatsen91 — APPROVED** [review]
Many thanks!
